_source/_posts/2021-05-28-using-okta-and-oso-to-secure-a-fastapi-and-sqlalchemy-app.md

```diff
@@ -0,0 +1,496 @@
+-from app.models import User
++from app.db import engine, setup_db, Base
++from app.models import User, Species
+ from app.schemas import Bear, BearBase
+ from app.seed import seed_db
+
+ # Load environment variables.
+ conf = Config(".env")
+ issuer, audience, client_id = conf("ISSUER"), conf("AUDIENCE"), conf("CLIENT_ID")
+
++# Initialize Oso.
++oso = Oso()
++register_models(oso, Base)
++oso.register_class(BearBase)
++oso.register_constant(Species, "Species")
++oso.load_file("app/policy.polar")
++
+
+@@ -46,2 +55,13 @@ def current_user(
+
++def get_authorized_db(request: Request):
++    get_oso = lambda: oso
++    get_user = lambda: request.state.user
++    get_action = lambda: request.scope["endpoint"].__name__
++    db = authorized_sessionmaker(get_oso, get_user, get_action, bind=engine)()
++    try:
++        yield db
++    finally:
++        db.close()
++
++
+ # Reset and seed database.
+@@ -62,8 +82,10 @@ app.add_middleware(
+ @app.get("/bears", response_model=List[Bear])
+-def index(db: Session = Depends(get_db)):
++def index(db: Session = Depends(get_authorized_db)):
+     return list_bears(db)
+
+
+ @app.post("/bears", response_model=Bear)
+ def create(request: Request, bear: BearBase, db: Session = Depends(get_db)):
++    if not oso.is_allowed(request.state.user, "create", bear):
++        raise HTTPException(403)
+     return create_bear(db, bear, request.state.user)
+diff --git a/app/policy.polar b/app/policy.polar
+new file mode 100644
+index 0000000..b458ddc
+--- /dev/null
++++ b/app/policy.polar
+@@ -0,0 +1,9 @@
++allow(user: User, "create", _: BearBase) if
++    not user.is_banned;
++
++allow(_: User, "index", _: User);
++
++allow(user: User, "index", bear: Bear) if
++    bear.owner = user or
++    bear.species in [Species.panda, Species.polar] or
++    bear.name = "Smokey";
+```
+
+## Summary
+
+We started out with a very fast and SQL-y application built on FastAPI and SQLAlchemy. We created and configured a new Okta application to handle identity management and authentication for our app. Then we used Oso to add efficient, fine-grained authorization to our back end API.
+
+The full example is available on [GitHub][].
+
+If you're up for it, here are a couple exercises for you, dear reader:
+
+- Add roles to the app using `sqlalchemy-oso`'s [built-in roles][] feature.
+  Perhaps every bear lives in a sanctuary, and a user can have a particular
+  role in each sanctuary, e.g., "Visitor", "Friend", or "Shepherd".
+- Add a "Delete" button next to every bear in the list and, wire it up to send
+  a DELETE request to the back end. Secure your new `delete()` handler with
+  Oso, and add a rule to the policy that only allows users to delete their own
+  bears.
+- Join [our slack][slack] and let us know what your favorite bear is and why
+  it's the polar bear.
+- Learn more about other use cases of Oso, other languages that it supports,
+  and more in [the Oso documentation][docs.osohq.com].
+
+[FastAPI]: https://fastapi.tiangolo.com/
```

Comment:
Add links to related blog posts and CTA (call-to-action) for following us (and Oso) on Twitter, etc.

Reply:
@jpf this is Oso's twitter handle: https://twitter.com/osoHQ